ethereum/core/src/main/java/tech/pegasys/teku/core/operationvalidators/AttesterSlashingStateTransitionValidator.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.core.operationvalidators;
+
+import static java.lang.Math.toIntExact;
+import static tech.pegasys.teku.core.operationvalidators.OperationInvalidReason.check;
+import static tech.pegasys.teku.core.operationvalidators.OperationInvalidReason.firstOf;
+import static tech.pegasys.teku.datastructures.util.AttestationUtil.is_slashable_attestation_data;
+import static tech.pegasys.teku.datastructures.util.AttestationUtil.is_valid_indexed_attestation;
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.get_current_epoch;
+import static tech.pegasys.teku.datastructures.util.ValidatorsUtil.is_slashable_validator;
+
+import com.google.common.collect.Sets;
+import com.google.common.primitives.UnsignedLong;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeSet;
+import tech.pegasys.teku.datastructures.operations.AttesterSlashing;
+import tech.pegasys.teku.datastructures.operations.IndexedAttestation;
+import tech.pegasys.teku.datastructures.state.BeaconState;
+
+public class AttesterSlashingStateTransitionValidator {
+
+  private List<UnsignedLong> indicesToSlash;
+
+  public Optional<OperationInvalidReason> validateSlashing(
+      final BeaconState state,
+      final AttesterSlashing attesterSlashing,
+      final Optional<Set<UnsignedLong>> intersectingIndices) {
```

Comment:
I'd probably make this method private and have a second public method that accepts a non-optional set of precalculated indices.

```suggestion
  public Optional<OperationInvalidReason> validateSlashing(
      final BeaconState state,
      final AttesterSlashing attesterSlashing,
      final Set<UnsignedLong> intersectingIndices) {
         return validateSlashing(state, attesterSlashing, Optional.of(intersectingIndices));
   }

  private Optional<OperationInvalidReason> validateSlashing(
      final BeaconState state,
      final AttesterSlashing attesterSlashing,
      final Optional<Set<UnsignedLong>> intersectingIndices) {
```

Reply:
Makes sense. Done.